### ISSUE TITLE
Proposal: Add the host's public key to the contract header

### ISSUE THREAD

**Issue by lukechampine** [lukechampine/us#3]
Currently the contract header only stores the contract's ID and secret key; the host key is stored as part of the revision. This means that if the revision becomes corrupted, the host key may be lost. Contract files are supposed to remain recoverable even if the revision is corrupted (indeed, they exploit this property in order to avoid excessive `fsync`ing). The header alone must be sufficient to recover the contract; thus, the host key must be moved into the header.

This will be an interesting case study in changing one of the `us` formats post-release. Previously I could change things with impunity, but now I have to maintain at least a semblance of compatibility.

## Implementation plan

The host key will be encoded as a `crypto.PublicKey` (a 32-byte array), and will immediately precede the contract ID in the header. That is, the new struct will be:
```go
type ContractHeader struct {
	magic   string
	version uint8
	hostKey crypto.PublicKey
	id      types.FileContractID
	key     crypto.SecretKey
}
```
This precludes non-ed25519 host keys, but that's fine: hosts are going to use ed25519 and ed25519 alone for the conceivable future. If circumstances change, we'll just have to change the format again.

While we're at it, I think we could encode the renter's secret key as just 32 bytes instead of 64. IIRC, `crypto.SecretKey` actually contains both the secret key *and* the public key, and the public key can be regenerated from the secret key. So the actual size of the header will not change.

When creating a contract, the `hostKey` field will be set in the obvious manner. After creation, the `hostKey` becomes immutable: those bytes will never be touched by another `Write` call.

When loading a contract, the revision will be checked against `hostKey` to ensure the keys match. A mismatch results in an error. (At some point, I will need to add helpers for restoring a revision after it's been corrupted, but that's outside the scope of this proposal.)

A new method, `(renter.Contract).HostKey`, will return the `hostKey` field as a `hostdb.HostPublicKey`. (This method will shadow the `HostKey` method of the `proto.ContractRevision` embedded in the `renter.Contract` struct.) This should result in all existing code transparently using the header key instead of the revision key.

The version number will be incremented to 2. `renter.LoadContract` will return an error if passed a contract with a version number other than 2.

A new function, `ConvertContractV1V2`, will be added to the `renter` package, along with testing.

A new command, `user contracts convert [filename]`, will be added to `user`. This command will upgrade a contract from v1 to v2. To prevent corruption, the upgraded contract will be written to a new file and then atomically `Rename`'d to overwrite the old file. The user is expected to run this command on all of their contracts. As a convenience, perhaps `user contracts convert` (with no argument) will automatically convert all of the user's available contracts. To point users in the right direction, `user` will suggest the `convert` command if it encounters a version mismatch error when loading a contract.

`ConvertContractV1V2` will be deleted two months after release, along with any associated testing and `user` code. (Future conversion functions will have longer lifespans; this one is short-lived because `us` does not have many users yet.) I do anticipate that `user contracts convert` will be necessary again someday; however, it will live eternally in version control, so I see no reason to keep it around while the format is stable.

In summary, the end goal is for users to get an error the next time they run `user`, and then run `user contracts convert` to upgrade all of their contracts to the new format.

**Comment by lukechampine** [comment]
Further notes on "is this really necessary?"

First, we could theoretically get the host key by scanning the blockchain for the original contract ID. But for various reasons, this sucks. If the contract never hit the blockchain, you're out of luck; also, the original contract only contains the *hash* of the unlock conditions, so you really need a contract *revision* (which reveals the host's public key), not the original contract. And if things are operating normally, there won't be any revisions on the chain until the contract is about to end.

Second, a more technical consideration: is the host key (contained in the revision) really in danger of being corrupted? Even though we're overwriting the revision, we're overwriting the host key with the *same bytes* each time. (The position of the host key within the file does not change.) So shouldn't it be impossible to corrupt the key?

I tried researching the semantics of `write`/`pwrite` in this case, but couldn't find anything conclusive. My gut says that in practice, there's no danger of the host key being corrupted. But my gut also says that, in the absence of specified behavior, we should assume that some filesystem out there will corrupt the key. So in the interest of safety, the most responsible thing to do is to put the key in the header, where we can be confident it will never be touched by another `write` call.
